### PR TITLE
#502 GzipInput and GzipOutput

### PR DESCRIPTION
**Pull request by fabriciofx**
As per #502.

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256 please, pay attention to this pull request

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos#503` is in scope.

**Comment by yegor256** [comment]
@fabriciofx sounds like a good idea, but it seems that it would be logical to gzip Input and Output instead. Don't you think?

**Comment by fabriciofx** [comment]
@yegor256 Sounds good. I'll resubmit a new PR. Thanks for your attention!

**Comment by fabriciofx** [comment]
@yegor256 Could you review this PR or assigned a REV, please?

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
The job [#503](https://github.com/yegor256/cactoos/issues/503) assigned to icok. The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by iCok** [comment]
@yegor256 There is problem with merge, I believe pull request owner can fix that, so what I supposed to do here?

**Comment by iCok** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/yegor256/cactoos/issues/503#issuecomment-357526354))

@iCok [@icok](http://www.0crat.com/u/icok) resigned from [#503](https://github.com/yegor256/cactoos/issues/503), please stop working

**Comment by fabriciofx** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/503#issuecomment-357675315))

@fabriciofx This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#503](https://github.com/yegor256/cactoos/issues/503) is in scope
  * The job is not assigned to anyone
  * These users are banned: [@fabriciofx](http://www.0crat.com/u/fabriciofx): This user reported the ticket; [@icok](http://www.0crat.com/u/icok): User was resigned from the ticket

**Comment by fabriciofx** [comment]
@yegor256 Could you assign a reviewer to this PR? Thanks!

**Comment by fabriciofx** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/503#issuecomment-358112413))

@fabriciofx This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#503](https://github.com/yegor256/cactoos/issues/503) is in scope
  * The job is not assigned to anyone
  * These users are banned: @fabriciofx[/z](http://www.0crat.com/u/fabriciofx)[/z](http://www.0crat.com/u/fabriciofx): This user reported the ticket; @icok[/z](http://www.0crat.com/u/icok)[/z](http://www.0crat.com/u/icok): User was resigned from the ticket

**Comment by fabriciofx** [comment]
@yegor256 Could you assign a REV to this PR? Thanks.

**Comment by fabriciofx** [comment]
@yegor256 Up

**Comment by yegor256** [comment]
@0crat assign @amihaiemil

File: src/main/java/org/cactoos/io/GzipInput.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+     */
+    private final int size;
+
+    /**
+     * Ctor.
+     * @param input The input.
+     */
+    public GzipInput(final Input input) {
+        // @checkstyle MagicNumberCheck (1 line)
+        this(input, 16 << 10);
```
Comment (amihaiemil) [inline src/main/java/org/cactoos/io/GzipInput.java]: @fabriciofx What does ``16 << 10`` mean? Could you write a plain ``int`` so it will be easier to understand?

File: src/test/java/org/cactoos/io/GzipInputTest.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            (byte) (GZIPInputStream.GZIP_MAGIC >> 8), (byte) 0x08,
+            (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00,
+            (byte) 0x00, (byte) 0x00, (byte) 0xF3, (byte) 0x48, (byte) 0xCD,
+            (byte) 0xC9, (byte) 0xC9, (byte) 0x57, (byte) 0x04, (byte) 0x00,
+            (byte) 0x56, (byte) 0xCC, (byte) 0x2A, (byte) 0x9D, (byte) 0x06,
+            (byte) 0x00, (byte) 0x00, (byte) 0x00,
+        };
+        MatcherAssert.assertThat(
+            "Can't read from a gzip input",
+            new GzipInput(
```
Comment (amihaiemil) [inline src/test/java/org/cactoos/io/GzipInputTest.java]: @fabriciofx I think everything could fit on a single line here: ``new GzipInput(...)``

File: src/test/java/org/cactoos/io/GzipOutputTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            (byte) 0x00, (byte) 0x00, (byte) 0xF3, (byte) 0x48, (byte) 0xCD,
+            (byte) 0xC9, (byte) 0xC9, (byte) 0x57, (byte) 0x04, (byte) 0x00,
+            (byte) 0x56, (byte) 0xCC, (byte) 0x2A, (byte) 0x9D, (byte) 0x06,
+            (byte) 0x00, (byte) 0x00, (byte) 0x00,
+        };
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        new LengthOf(
+            new TeeInput(
+                "Hello!",
+                new GzipOutput(
```
Comment (amihaiemil) [inline src/test/java/org/cactoos/io/GzipOutputTest.java]: @fabriciofx same line here as well

File: src/test/java/org/cactoos/io/GzipInputTest.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import org.junit.Test;
+
+/**
+ * Test case for {@link org.cactoos.io.GzipInput}.
+ * @author Fabricio Cabral (fabriciofx@gmail.com)
+ * @version $Id$
+ * @since 0.29
+ * @checkstyle JavadocMethodCheck (500 lines)
+ */
+public final class GzipInputTest {
```
Comment (amihaiemil) [inline src/test/java/org/cactoos/io/GzipInputTest.java]: @fabriciofx Maybe add a unit test for closed and one for empty stream? Same for output.

File: src/main/java/org/cactoos/io/GzipInput.java (hunk 1 above)
Comment (fabriciofx) [reply src/main/java/org/cactoos/io/GzipInput.java]: @amihaiemil It's the buffer's size. I just put the same value used by Yegor [here](https://github.com/yegor256/cactoos/blob/master/src/main/java/org/cactoos/io/InputAsBytes.java#L61).

File: src/test/java/org/cactoos/io/GzipOutputTest.java (hunk 3 above)
Comment (fabriciofx) [reply src/test/java/org/cactoos/io/GzipOutputTest.java]: @amihaiemil Ok. I'll change it.

File: src/main/java/org/cactoos/io/GzipInput.java (hunk 1 above)
Comment (amihaiemil) [reply src/main/java/org/cactoos/io/GzipInput.java]: @fabriciofx I see, but I would still go with an integer -- what does ``16 << 10`` evaluate to, I don't know, and I bet I am not the only one :D.  ``<<`` is one of the strangest operators in Java, I have never used it in 4 years so far

**Comment by 0crat** [comment]
> @0crat assign @amihaiemil ([here](https://github.com/yegor256/cactoos/issues/503#issuecomment-358948437))

@yegor256 The job [#503](https://github.com/yegor256/cactoos/issues/503) assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://datum.zerocracy.com/pages/policy.html#19): -5 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+7695](https://www.0crat.com/u/yegor256)

**Comment by amihaiemil** [comment]
@fabriciofx what is the status here, did you have time to do those changes?

**Comment by amihaiemil** [comment]
@fabriciofx ping

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by amihaiemil** [comment]
@fabriciofx what about this one? Can you do the changes so we can merge it?

**Comment by fabriciofx** [comment]
@amihaiemil My changes are almost done. I'll submit in few hours.

File: src/main/java/org/cactoos/io/GzipInput.java (hunk 1 above)
Comment (rexim) [reply src/main/java/org/cactoos/io/GzipInput.java]: @amihaiemil As far as I know, `2 << n` is a super quick way to rise `2` to the power of `n`. `16` is a power of `2`: `16 == 2⁴`. So `16 << 10 == 2¹⁴` (check it). Programmers have an old tradition of allocating buffers with the size of the power of `2`. I don't know why. Probably somehow related to the memory alignment or something. :)

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 8 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by amihaiemil** [comment]
@fabriciofx how are we here? DId you do the changes? This PR should be merged soon as 0crat says ^^

**Comment by 0crat** [comment]
The user @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) resigned from [#503](https://github.com/yegor256/cactoos/issues/503), please stop working

**Comment by amihaiemil** [comment]
@yegor256 You said [here](https://github.com/zerocracy/farm/issues/463) that there is no deadline for the CR, that it was a bug. Either it fixed but the fix doesn't apply retroactively or the bug is still there :)

**Comment by yegor256** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/503#issuecomment-361582183))

@yegor256 This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#503](https://github.com/yegor256/cactoos/issues/503) is [in scope](http://datum.zerocracy.com/pages/policy.html#14) for 1mon
  * The role is `DEV`
  * The job is not assigned to anyone
  * The [budget](http://datum.zerocracy.com/pages/policy.html#4) is 30 minutes
  * These users are banned and won't be assigned:
    * @fabriciofx[/z](https://www.0crat.com/u/fabriciofx): This user reported the ticket
    * @icok[/z](https://www.0crat.com/u/icok): User was resigned from the ticket
    * @amihaiemil[/z](https://www.0crat.com/u/amihaiemil): User was resigned from the ticket

**Comment by yegor256** [comment]
@amihaiemil the problem is that the role here is `DEV` (old bug, fixed already).

**Comment by yegor256** [comment]
@0crat assign @amihaiemil

**Comment by 0crat** [comment]
> @0crat assign @amihaiemil  ([here](https://github.com/yegor256/cactoos/issues/503#issuecomment-361582403))

@yegor256 @yegor256[/z](https://www.0crat.com/u/yegor256) not enough funds available in the project, can't set budget of job [#503](https://github.com/yegor256/cactoos/issues/503), see [§21](http://datum.zerocracy.com/pages/policy.html#21)

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://datum.zerocracy.com/pages/policy.html#19): -5 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+8955](https://www.0crat.com/u/yegor256)

**Comment by amihaiemil** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/503#issuecomment-361584085))

@amihaiemil This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#503](https://github.com/yegor256/cactoos/issues/503) is [in scope](http://datum.zerocracy.com/pages/policy.html#14) for 1mon
  * The role is `DEV`
  * The job is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) for [7min](http://datum.zerocracy.com/pages/policy.html#8)
  * There is no monetary reward attached, it's a [free](http://datum.zerocracy.com/pages/policy.html#2) job
  * The job doesn't have any [impediments](http://datum.zerocracy.com/pages/policy.html#9)
  * The [budget](http://datum.zerocracy.com/pages/policy.html#4) is 30 minutes
  * These users are banned and won't be assigned:
    * @fabriciofx[/z](https://www.0crat.com/u/fabriciofx): This user reported the ticket
    * @icok[/z](https://www.0crat.com/u/icok): User was resigned from the ticket
    * @amihaiemil[/z](https://www.0crat.com/u/amihaiemil): User was resigned from the ticket

**Comment by amihaiemil** [comment]
@fabriciofx ping?

**Comment by yegor256** [comment]
@0crat assign @amihaiemil

**Comment by 0crat** [comment]
> @0crat assign @amihaiemil  ([here](https://github.com/yegor256/cactoos/issues/503#issuecomment-361584850))

@yegor256 Job [#503](https://github.com/yegor256/cactoos/issues/503) is already assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Order was successfully finished: +30 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), total is [+450](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://datum.zerocracy.com/pages/policy.html#28): +15 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+9595](https://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
The job [#503](https://github.com/yegor256/cactoos/issues/503) is now out of scope
